File: Justaway/src/main/java/info/justaway/UserSearchActivity.java
package info.justaway;

import android.app.ActionBar;
import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.FragmentActivity;
import android.util.Log;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.view.View;
import android.widget.AbsListView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.ProgressBar;

import info.justaway.adapter.UserAdapter;
import twitter4j.ResponseList;
import twitter4j.User;

public class UserSearchActivity extends FragmentActivity {

    private EditText mSearchWords;
    private int mPage = 1;
    private UserAdapter mAdapter;
    private ListView mListView;
    private ProgressBar mProgressBar;
    private boolean mAutoLoading = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        JustawayApplication.getApplication().setTheme(this);
        setContentView(R.layout.activity_user_search);

        ActionBar actionBar = getActionBar();
        if (actionBar != null) {
            actionBar.setHomeButtonEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        Button search = (Button) findViewById(R.id.search);
        Typeface fontello = JustawayApplication.getFontello();
        search.setTypeface(fontello);

        mSearchWords = (EditText) findViewById(R.id.searchWords);
        mProgressBar = (ProgressBar) findViewById(R.id.guruguru);
        mListView = (ListView) findViewById(R.id.list_view);
        mListView.setVisibility(View.GONE);

        // ユーザをViewに描写するアダプター
        mAdapter = new UserAdapter(this, R.layout.row_user);
        mListView.setAdapter(mAdapter);

        search.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                search();
            }
        });

        mSearchWords.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                //EnterKeyが押されたかを判定
                if (event.getAction() == KeyEvent.ACTION_DOWN
                        && keyCode == KeyEvent.KEYCODE_ENTER) {
                    search();
                    return true;
                }
                return false;
            }
        });

        Intent intent = getIntent();
        String query = intent.getStringExtra("query");
        if (query != null) {
            mSearchWords.setText(query);
            search.performClick();
        } else {
            JustawayApplication.getApplication().showKeyboard(mSearchWords);
        }

        mListView.setOnScrollListener(new AbsListView.OnScrollListener() {

            @Override
            public void onScrollStateChanged(AbsListView view, int scrollState) {
            }

            @Override
            public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
                // 最後までスクロールされたかどうかの判定
                if (totalItemCount == firstVisibleItem + visibleItemCount) {
                    additionalReading();
                }
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return true;
    }

    private void additionalReading() {
        if (!mAutoLoading) {
            return;
        }
        mProgressBar.setVisibility(View.VISIBLE);
        mAutoLoading = false;
        new UserSearchTask().execute(mSearchWords.getText().toString());
    }

    private void search() {
        JustawayApplication.getApplication().hideKeyboard(mSearchWords);
        if (mSearchWords.getText() == null) return;
        mAdapter.clear();
        mListView.setVisibility(View.GONE);
        mProgressBar.setVisibility(View.VISIBLE);
        new UserSearchTask().execute(mSearchWords.getText().toString());
    }

    private class UserSearchTask extends AsyncTask<String, Void, ResponseList<User>> {
        @Override
        protected ResponseList<User> doInBackground(String... params) {
            String query = params[0];
            try {
                Log.d("justaway", "task");
                return JustawayApplication.getApplication().getTwitter().searchUsers(query, mPage);
            } catch (Exception e) {
                e.printStackTrace();
                return null;
            }
        }

        @Override
        protected void onPostExecute(ResponseList<User> users) {
            mProgressBar.setVisibility(View.GONE);
            if (users == null) {
                JustawayApplication.showToast(R.string.toast_load_data_failure);
                return;
            }
            for (User user : users) {
                mAdapter.add(user);
            }
            if (users.size() == 20) {
                mAutoLoading = true;
                mPage++;
            }
            mListView.setVisibility(View.VISIBLE);
        }
    }
}

File: Juktaway/src/main/java/net/slash_omega/juktaway/UserSearchActivity.kt
package net.slash_omega.juktaway

import android.os.Bundle
import android.view.KeyEvent
import android.view.MenuItem
import android.view.View
import android.widget.AbsListView
import jp.nephy.penicillin.endpoints.users
import jp.nephy.penicillin.endpoints.users.search
import jp.nephy.penicillin.extensions.await
import kotlinx.android.synthetic.main.activity_user_search.*
import kotlinx.coroutines.launch
import net.slash_omega.juktaway.adapter.UserAdapter
import net.slash_omega.juktaway.twitter.currentClient
import net.slash_omega.juktaway.util.KeyboardUtil
import net.slash_omega.juktaway.util.ThemeUtil
import org.jetbrains.anko.toast

/**
 * Created on 2018/08/29.
 */
class UserSearchActivity: ScopedFragmentActivity() {
    private lateinit var mSearchWord: String
    private var mPage = 1
    private lateinit var mAdapter: UserAdapter
    private var mAutoLoading = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ThemeUtil.setTheme(this)
        setContentView(R.layout.activity_user_search)

        actionBar?.run {
            setHomeButtonEnabled(true)
            setDisplayHomeAsUpEnabled(true)
        }

        list_view.visibility = View.GONE
        mAdapter = UserAdapter(this, R.layout.row_user)
        list_view.adapter = mAdapter

        search.setOnClickListener { search() }
        search_text.setOnKeyListener { _, code, e ->
            if (e.action == KeyEvent.ACTION_DOWN && code == KeyEvent.KEYCODE_ENTER) {
                search()
                true
            } else false
        }

        intent.getStringExtra("query")?.let {
            search_text.setText(it)
            search.performClick()
        } ?: KeyboardUtil.showKeyboard(search_text)

        list_view.setOnScrollListener(object: AbsListView.OnScrollListener {
            override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {}

            override fun onScroll(view: AbsListView, firstVisibleItem: Int, visibleItemCount: Int, totalItemCount: Int) {
                // 最後までスクロールされたかどうかの判定
                if (totalItemCount == firstVisibleItem + visibleItemCount) {
                    additionalReading()
                }
            }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }
        return true
    }

    private fun search() {
        KeyboardUtil.hideKeyboard(search_text)
        if (search_text.text == null) return
        mAdapter.clear()
        mPage = 1
        list_view.visibility = View.GONE
        guruguru.visibility = View.VISIBLE
        mSearchWord = search_text.text.toString()
        userSearch(mSearchWord)
    }

    private fun additionalReading() {
        if (!mAutoLoading) return
        guruguru.visibility = View.VISIBLE
        mAutoLoading = false
        userSearch(mSearchWord)
    }

    private fun userSearch(word: String) = launch {
        runCatching {
            currentClient.users.search(word, mPage, 20).await()
        }.onSuccess { res ->
            mAdapter.addAll(res)
            if (res.size == 20) {
                mAutoLoading = true
                mPage++
            }
            list_view.visibility = View.VISIBLE
        }.onFailure {
            toast(R.string.toast_load_data_failure)
        }

        guruguru.visibility = View.GONE
    }
}